### ISSUE TITLE
3.x.x: Pool-acquire timing covers transactions only; non-transactional acquisitions are unmeasured

### ISSUE THREAD

**Issue by kiwidream** [Qbit-Org/qbit-mining-bootstrap#352]
**Completed — 2026-09-18:** #457 merged as `d78659de608ad30645ba630f15eb8be998b3e42a`. The consolidated census, explicitly allowlisted exclusions, partition-maintenance and remaining coordinator checkout timing, CI guard/negative controls, and matching documentation/outcome tests satisfy the four agreed closing conditions below. Final review: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/457#issuecomment-5734174173 . All 19 CI checks passed, including the repaired API fixture isolation regression. Public-role export remains explicitly excluded, not silently claimed covered.

**Agreed finishing direction:** consolidate the remaining work into one bounded PR rather than further per-method slices. Done when (1) the current run-role source census identifies every fresh acquisition as timed or explicitly allowlisted with reason/owner, distinguishing borrowed connections from fresh checkouts; (2) uncovered partition-maintenance acquisition reachable from the share path is timed without changing SQL, retry or connection boundaries; (3) a CI source guard detects unclassified new acquisition sites, with negative controls; and (4) the coverage document and outcome/cancellation tests match that population. Startup/migration/operator pools must be explicitly listed, not silently counted as covered. Public-role export remains a separate, explicit policy decision; its exclusion is not a claim of public-pool coverage. Any other live run-role gaps found by the final census belong in the same finishing scope or must be explicitly dispositioned before closure.

The counts below describe the historical #334 baseline, not the current source census.

---

Raised against #334 (#328) from the #265 slice 3 review. Not a defect in #334, whose documented scope already says so, but a real coverage gap worth closing deliberately.

## What is measured today

`qbit_prism_database_pool_acquire_seconds` is recorded in two places:

- the metrics collector's own `pool.acquire()`;
- the ledger's timed `begin()` helper, so every ledger **transaction**.

A query run straight against the pool — `fetch_one(&self.pool)`, `execute(&self.pool)` and friends — acquires a connection implicitly through sqlx's `Executor for &Pool` and is never timed.

Counted on the #334 branch, excluding tests: **36** timed transaction sites against **29** untimed direct-pool sites, spread across `ledger/{connect,audit,migration,jobs,fanout,difficulty,blocks}.rs`, `coordinator.rs`, `coordinator/{miner_submit,work_ledger}.rs` and `api/public_service.rs`.

## What this does and does not mean

- **The wait signal is still representative for the `run` role.** An untimed read consumes a permit from the same pool, so a timed transactional acquisition queues behind it and records that wait. Under saturation the p99 rises, and the alert's "at least ten completed acquisitions in five minutes" guard is met by share appends and the collector alone. The alarm does not go quiet.
- **The count is a sample, not a census.** Roughly a third of acquisitions are missing, so `_count` must not be read as pool throughput or as a request rate.
- **The public role has no coverage at all.** `api/public_service.rs` opens no transactions, and this family is a `run`-role family, so pool pressure in the public read service is not measured by anything today. That is the sharper end of this gap.

## Proposed change

Add a timed acquire helper beside the existing `begin()` — same `WaitGuard`, same `success`/`failure` and cancellation semantics — and route the non-transactional sites through it, so the family counts every acquisition the process makes. Then decide separately whether the public role should export the family.

Touching 29 sites across three owners' areas is why it was not folded into #334.

Refs #328, #334.

**Comment by kiwidream** [comment]
#353 has merged, and it grows the untimed set in the most acquisition-heavy way yet. Counted on the rebased #334 branch, the untimed direct-pool sites went from 29 to 31.

The new path is the durable-range proof that now runs **before** the settlement lock in `Ledger::land_candidate_checked`:

- a `fetch_one(&self.pool)` probe for whether the block already has an audit row;
- when it does not, `verify_durable_range(&self.pool, …)` in `ledger/audit.rs`, which loops `fetch_all(pool)` one page at a time with `VERIFY_PAGE_ROWS = 4096` until the whole window is matched.

So landing one block reads the entire window off the pool in pages, roughly thirty acquisitions for a 125,000-share window, and none of them are timed. The transaction that follows is timed, and records only its own wait.

That makes the count gap concrete rather than theoretical: during a found-block landing the family records one acquisition where about thirty occurred, and the unrecorded ones are the largest reads in the system. The wait signal is still sound, for the reason in the issue body — these reads hold permits that the following timed acquisition then queues for — but any reading of `_count` as acquisition volume is off by the paging factor exactly when a block lands.

Credit for spotting this before #353 merged goes to the #265 slice 3 review.

**Comment by djh58** [comment]
@kiwidream I can take a bounded B contribution here in parallel with #273. Proposed split: prepare the canonical timed-acquire helper using the existing wait guard and success/failure/cancellation semantics, then route only agreed B-owned non-transactional call sites through it. Please confirm the helper/export seam and whether you already have an implementation in flight. I will leave A/C-owned ledger callers and the separate public-role metric decision untouched until ownership is agreed. A worker is inventorying the current native sites and preparing the bounded patch plan; no competing cross-owner edits are authorized.

**Comment by djh58** [comment]
Taking implementation of #352, assigned to djh58, starting from current 3.x.x. We will add the shared timed-acquire helper and convert non-transactional callers with success/failure/cancellation accounting while preserving checkout lifetimes, existing deadlines and query behavior. The public-role export policy remains a separate decision as scoped above. I will coordinate overlapping #273/#350 files with their active authors. @kiwidream please flag any existing competing helper implementation; we are proceeding with the helper and independent caller batches meanwhile.

**Comment by djh58** [comment]
Follow-up from the completed review of #368 at `bed69c7`: consolidate the acquisition observers before expanding the remaining #352 callers.

`metrics/collectors.rs` currently uses `PoolAcquireObservation` with `tokio::time::Instant`; `ledger/connect/acquire.rs` uses the existing `WaitGuard` with `std::time::Instant`. Both preserve acquisition outcomes today, but two implementations can drift. #368 retains both existing clocks; consolidation belongs in the next #352 slice, with an explicit shared clock decision and compatibility tests for the collector's paused-time tests and real elapsed waits.

Acceptance boundaries for that follow-up:

- Share one acquisition-observation implementation without adding a third observer. Arm on first poll; an unpolled future records nothing; absent metrics read no clock; success, acquisition error and polled cancellation each record exactly once.
- Preserve returned values/errors, cancellation cleanup, pool/connection lifetimes and existing deadlines. Stop timing before SQL or `BEGIN`; do not turn query failures into checkout failures or record existing-transaction `Acquire` calls as fresh pool checkouts.
- Keep exact isolated count/sum tests, paused-clock compatibility tests, real exhausted-pool cancellation/reuse, and one-slot subsequent-transaction coverage. Test the delayed/failed/cancelled `BEGIN` boundary explicitly when extending that coverage.

The remaining integration inventory includes job/coordinator reads; audit paging, block reconciliation and migration-source/audit-export reads; startup validation, session cleanup, tools and rollups. Preserve per-page checkout release and coordinate `ledger/jobs.rs` / `coordinator/work_ledger.rs` overlaps with #363 and the #273 repair work in #365 before edits. API/read-pool instrumentation and public-role export policy remain separate decisions. #368 adds eight selected callers and transaction-helper reuse; it does not complete #352 or make `_count` an acquisition census or request rate.

**Comment by djh58** [comment]
PR #368 merged into 3.x.x as d733ef9 after full review disposition and final-head CI: the shared timed-acquire helper and eight selected direct callers are landed, with first-poll/cancellation/exactly-once tests and 272 required integration gates executed. Remaining caller integrations, observer consolidation with clock-compatible tests, and the separately considered public-role export policy remain open here. This is partial coverage, not a complete acquisition census.

**Comment by djh58** [comment]
PR #376 merged into 3.x.x as 387064c0ebc1478e461a6b8a32888a390d8cf885. Ledger and collector acquisition timing now share one implementation, preserving first-poll timing, exactly-once outcomes, cancellation cleanup, and checkout-before-SQL boundaries. Initial medium/Codex/Fable/thermo reviews are accounted for; thermo fixes received independent final-head verification at 4689b9e, and final CI proved all 272 required gates executed. The diagnostic latency retry and unresolved original cause remain documented on the PR. This issue stays open: remaining direct-pool callers and public-role coverage/export policy are not completed by this slice.

**Comment by djh58** [comment]
PR #390 merged into 3.x.x as 553fa356e93b46c4226bb667ed4321ec1dc5b82a. Seven audit, landing/reconciliation and migration checkout sites now use the shared observer, preserving first-poll timing, exactly-once cancellation outcomes, SQL/error behavior and release before blocking work. Complete review ledger: 4 fixed, 2 rejected, 0 deferred, 0 open; final qualification 432 local tests, 185/185 local gates and 291/291 hosted gates, all 20 checks including exact-head Cursor green. This is partial coverage: audit reconstruction, coordinator/job/candidate/startup/session callers, rollup transactions, inherited boundary probes and the separately considered public-role export policy remain. Issue stays open and board stays In progress.

**Comment by djh58** [comment]
PR #394 merged into 3.x.x as 7ec7d3af41c23a32144bb2a38239102f9e07d546. Ledger::job, compact_prepared and prune_expired_jobs now observe connection checkout through the shared timer without changing SQL, errors or deadlines. Four initial review lanes plus independent final test-only delta verification completed: 3 fixed, 3 rebutted, none deferred/open; all 19 CI workflow checks and exact-head Cursor passed. Final ledger: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/394#issuecomment-5687649654 . This issue remains open/In progress for the documented remaining direct-pool callers and separately considered public-role export policy.

**Comment by Anatolie** [comment]
#369 (merged on 3.x.x as `5af6519c`) adds one non-transactional checkout site that is currently in neither list in `docs/prism-pool-acquire-metrics.md` — not the covered set, and not the remaining-untimed inventory. Flagging it so the census stays honest.

**The site.** `apply_online_migration` in `crates/qbit-prism-server/src/ledger/migration/online.rs`:

```rust
let mut connection = pool.acquire().await?.detach();
```

One checkout per pending online migration during `Ledger::connect`, immediately detached from the pool, then given session-level `statement_timeout=0` / `lock_timeout=0` and held for the whole `CREATE INDEX CONCURRENTLY` / `DROP INDEX CONCURRENTLY` sequence — hours on a production-sized ledger.

**Timing it is a one-liner.** The function already takes `metrics: Option<&crate::metrics::Metrics>`, so it is the `audit.rs` pattern:

```rust
let mut connection = crate::metrics::time_pool_acquire(metrics, pool.acquire()).await?.detach();
```

No signature change, no new import. (`Ledger::acquire()` is a `&self` method and this function takes a `&PgPool`, so it would use the underlying helper directly.)

**Whether it should be timed is a real judgment call**, which is why this is a note rather than a change. It is a genuine pool permit, so by the census goal it counts. But it is also once per process start, detached rather than returned, and held for hours — closer to the "operator-only connections" carve-out already in the doc than to the steady-state pressure the family is read for. Timing it adds roughly one observation per startup, which is negligible for percentiles but not nothing for a `_count` that is meant to become a census.

Either resolution is fine and both are one line; what should not persist is the current state, where the site appears in neither the covered list nor the untimed inventory. Worth folding into whichever slice of this issue next touches `ledger/migration`.

**Comment by djh58** [comment]
Starting a separate bounded follow-up on current 3.x.x for the online-migration checkout identified above. Scope: observe apply_online_migration connection acquisition through the existing timer before detach, add focused regression evidence, and update the covered/remaining inventory. The observation measures checkout only, not index-building time or the detached connection lifetime. No migration SQL, lock/timeout/cleanup semantics, public-role export policy, or runtime-activation changes are included. This will be its own Refs #352 PR with independent review; the broader issue remains open.

**Comment by djh58** [comment]
PR #403 merged into 3.x.x as 3cbee3e7118837c07e3d707fa5fd68b3e1982f3a. Online-migration pool checkout is now observed before detach, excluding subsequent session settings, SQL and detached lifetime. Deep-review complete: 4 fixed groups, 3 rejected items, no deferred/open findings; 20 checks passed and all 358 expected integration gates independently verified. Final ledger: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/403#issuecomment-5699373437 . The initial throughput miss and unchanged-head passing retry remain documented; no performance fix is claimed. #352 stays open for remaining direct-pool acquisition sites and the separately considered public-role export policy. Board returns to In progress.

**Comment by djh58** [comment]
Starting the next bounded slice on current 3.x.x after #403: instrument only the three independent bootstrap/newest/oldest boundary-probe checkouts in verify_durable_range. Existing paged acquisitions are already observed. SQL, proof decisions, error precedence, deadlines and release-before-blocking boundaries remain unchanged. Separate PR with focused PostgreSQL observation tests and deep-review; broader reconstruction, candidate, runtime and public-role policy work stays out of scope.

**Comment by djh58** [comment]
PR #417 merged into 3.x.x as e5962e508044d5b79694af0aface6e6ecf3a6cc6. The bootstrap existence, newest-boundary and partial-history oldest-boundary audit probes now use timed checkout, preserving SQL and statement-scoped release. Nine focused PostgreSQL contracts, negative control, 367 expected CI gates, and completed independent review support this slice. Final ledger: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/417#issuecomment-5701078628 . This issue stays open for remaining direct-pool acquisition coverage and the separate public-role export policy; the historical opening inventory is not a current coverage census.

**Comment by djh58** [comment]
PR #436 merged into 3.x.x as 58b004fc74bf23336a493ece0639092c207a9b55. Session reservation explicit release and background Drop now measure pool checkout waits before the unchanged token-fenced DELETE; SQL execution is excluded and cancellation/fallback semantics remain intact. Final deep-review: 1 documentation fix, 3 advisory dispositions, zero open defects or incomplete lanes; 19 final-head CI checks passed. Independent Fable verified the five new PostgreSQL tests plus a scratch cancellation/replacement-token probe. Final ledger: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/436#issuecomment-5716469994 . #352 remains open for the remaining acquisition sites and the separate public-role export policy; this is a landed slice, not complete coverage.

**Comment by djh58** [comment]
Starting the next bounded implementation slice after #436: time fresh pool checkouts in audit reconstruction (audit_canonical_bytes, materialize_audit_row snapshot lookup and range reads), preserving SQL, audit bytes/digests, error ordering, deadlines and original release boundaries. Existing borrowed connections and transactions will not be double-counted. Separate focused PostgreSQL observation/cancellation tests, a Refs #352 PR and deep-review; candidate, startup, public-role and rollup coverage remain outside this slice.

**Comment by djh58** [comment]
Merged #442 into 3.x.x as b51e8e898f6cee8e9c377c1c6f4ccd095f148be4 on 2026-09-17. This lands the audit-reconstruction checkout timing slice: native inline snapshot reads now observe two successful acquisitions and range-backed reads three; imported canonical/legacy/missing-row paths remain one. Connection release, bytes, digest checks and error ordering are preserved.

Deep-review complete at f258edb1: four findings fixed, five rebutted, no open/deferred findings; all 20 checks green, all 529 expected database gates executed. Final dispositions: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/442 .

#352 remains open/In progress for remaining direct-pool callers and the separately scoped public-role export policy; this merge does not claim complete acquisition coverage.

**Comment by djh58** [comment]
Starting the next bounded coordinator-owned timing slice after #442: WorkLedger::now_ms and the two direct checkouts in persist_block_only (initial duplicate check and credit/disposition poll). The implementation preserves SQL, the original acknowledgement deadline, candidate/credit outcomes and connection-release boundaries; it only routes acquisitions through the existing timing helper. Separate branch/PR against 3.x.x with focused PostgreSQL regressions and full independent review. Public-role metrics policy and unrelated candidate/startup sites remain outside this slice.

**Comment by djh58** [comment]
Merged #446 into 3.x.x as 0493c13c685b5440c5e6e6299fa589e0a2de831b. This lands timing for WorkLedger::now_ms, coherent chain-observation reads, and both block-only duplicate/disposition probes while preserving the incoming partition filters and original acknowledgement deadlines. Final review: 3 fixed, 4 rebutted, zero deferred/open; 19 checks passed and downloaded CI evidence proves all 591 gated tests executed. Ledger: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/446#issuecomment-5719509870 .

#352 remains open/In progress: this is not a claim of complete acquisition coverage. Next is a source census on the merged base to distinguish genuinely untimed fresh acquisitions from existing timed transactions/borrowed connections and list the separate public-role export decision.

**Comment by djh58** [comment]
PR #453 merged into 3.x.x as 4cb873213ec7828f4b2b4cb36be1f7caf65765d7. The run-role rollup loop now records checkout duration/outcome before BEGIN; SQL, COMMIT, watermark and public wrapper semantics remain unchanged. Four independent review lanes completed with no required findings; all 19 checks and 598 exact-head gates passed. #352 stays open: this is another bounded coverage slice, not full pool-throughput coverage or a public-role export-policy decision. The completed source census remains the basis for any next slice.

**Comment by djh58** [comment]
Merged #454 into 3.x.x as 587aef599611319400c28369501a89d0cae120b8. Landed-audit reads and bits updates now use the existing checkout timer; all 20 active checks passed with no required review defects. This remains partial coverage. The issue body now states the consolidated finishing contract: final run-role census, remaining partition-path timing, an allowlisted CI coverage guard, and honest documented exclusions. No additional finishing worker is claimed active yet.

**Comment by djh58** [comment]
The consolidated finishing worker is now active from merged #454/base 587aef59: final run-role acquisition census, remaining partition-path coverage, allowlisted CI source guard with negative controls, and coverage documentation. The issue remains In progress. This replaces further per-method instrumentation PRs; public-role export remains explicitly separate.

**Comment by djh58** [comment]
The consolidated final coverage candidate is published as PR #457, targeting 3.x.x at signed head 29cc9df8bdda003c2bd94014411722bd51ad20a2. It includes partition checkout timing, the two remaining candidate-heartbeat fallback acquisitions, and a checked source census with negative controls through existing CI discovery. Local qualification is documented in the PR. Independent focused review has started; hosted CI and review remain acceptance gates, and this issue stays open until those and the closing criteria are verified.

**Comment by djh58** [comment]
Completed by merged #457 (d78659de). All four agreed closing conditions are met: current acquisition census with explicit exclusions, partition and remaining coordinator timing, CI source guard/negative controls, and aligned coverage/outcome documentation/tests. Final review and CI evidence: https://github.com/Qbit-Org/qbit-mining-bootstrap/pull/457#issuecomment-5734174173 . This is not a claim that excluded public/operator pools are instrumented.
